### DISCUSSION TITLE
Calling a ContentWriter from the front end

### DISCUSSION THREAD

**Discussion by BubbaRichard** [MonoGame/MonoGame#8259] [Content]
Hello,
I've searched and I'm not finding :) an answer.
I want to use the Content Pipeline to mod my game also. I remember someone on the xna forums had an example but couldn't find it. I believe I need to get the content complier, declare my writer and pass it to a compile method but the documentation is a little lacking on the serialization side. And it doesn't look like that the correct way to go as the Complier takes the actual object to be complied and a stream, etc. Has anyone tried this and have a starting point? I know I fiddled with this back in xna 1 and 2 because I had the same idea then.
Thanks,
Bubba

**Comment by BlizzCrafter** [comment]
I'm struggeling a bit to understand what you are trying to achieve. 

Do you want to create a Content Pipeline Project for your own custom data or do you want to manually compile content from within your game/editor?

**Comment by BubbaRichard** [comment]
I have a custom pipeline, it works fine. If a player creates a model in blender, exports it like I tell them to I want to give them a way to compile it vs me having a second method to load mods. I could totally use a binary serializer reader and writer to do it myself but I'm just trying to figure a way to let players compile their mods to xnb files.

**Comment by BlizzCrafter** [comment]
> I'm just trying to figure a way to let players compile their mods to xnb files

There are different ways of doing this.

1. You could ship the MGCB files together with your game and then just call the MGCB.exe from your game or via batch file for example. You would need to dynamically create a .mgcb file through, but this is not a big deal. I already saw some games on Steam doing it like this (e.g. [Sweet Transit](https://steamdb.info/app/1612770/)).

or

2. You could create something like a SDK package for your game which contains a build script and when the user runs the build script, it would install the MGCB files locally for a specific mod project via `dotnet`. This approach has the advantage that you don't need to ship the MGCB file (and all of its dependencies) by yourself.

or

3. You could implement your own content writer like you said, but you would need a reference to the MonoGame.Framework.Content.Pipeline.dll and the dependencies for content compilation anyways.

I hope this gave you some ideas. 

Personally I would go for option 2 nowadays. It's the most safest way in my opinion. On the other hand it should be legally no problem to just ship the MGCB files together with your game and go for route 1. Route 3 is kind of unnecessary besides you want explicit fine control about your content compilation.

**Comment by BubbaRichard** [comment]
I don't believe Including MGCB is an option and with all parts needed there in the library, I'm just looking for the pattern to use on compiling from the front end. Thanks!

**Reply by BlizzCrafter** [reply]
Then [MonoGame.Framework.Content.Pipeline](https://github.com/MonoGame/MonoGame/tree/develop/MonoGame.Framework.Content.Pipeline) is your friend.

Take a look  at the [PipelineManager](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/Builder/PipelineManager.cs). Espacially when it comes to [building your content](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/Builder/PipelineManager.cs#L534).

I remember doing this myself for a project and I created the [MonoGame.RuntimeBuilder](https://github.com/BlizzCrafter/MonoGame.RuntimeBuilder) afterwards. However this is just a wrapper around the MGCB and it's an outdated project. But maybe this can give additional help when creating your own format. In my [Build](https://github.com/BlizzCrafter/MonoGame.RuntimeBuilder/blob/master/MonoGame.RuntimeBuilder/MGCB/Builder.cs#L192) method I made use of the [PipelineManager](https://github.com/BlizzCrafter/MonoGame.RuntimeBuilder/blob/master/MonoGame.RuntimeBuilder/MGCB/Builder.cs#L206) as well as you can see. This is now contained [here](https://github.com/MonoGame/MonoGame/tree/develop/MonoGame.Framework.Content.Pipeline/Builder) in the MonoGame repo.

I hope that this additional information is somewhat useful and helps you out.
